loader/loader.go

```diff
@@ -545,69 +547,38 @@ func (l *Loader) protoEnum(tspec *ast.TypeSpec) (*desc.EnumDescriptorProto, erro
 	return enum, nil
 }
 
-func (l *Loader) protoType(expr ast.Expr, pkg *types.Package) (desc.FieldDescriptorProto_Type, desc.FieldDescriptorProto_Label, string) {
-	if pkg == nil {
-		pkg = l.tpkg
-	}
-	switch x := expr.(type) {
-	case *ast.Ident:
+func (l *Loader) protoType(typ types.Type) (desc.FieldDescriptorProto_Type, desc.FieldDescriptorProto_Label, string) {
+	switch typ := typ.(type) {
+	case *types.Basic:
 		// Map Go types to proto types
 		// https://developers.google.com/protocol-buffers/docs/proto3#scalar
-		switch x.Name {
-		case "string":
-			return desc.FieldDescriptorProto_TYPE_STRING, 0, x.Name
-		case "int", "int32":
-			return desc.FieldDescriptorProto_TYPE_INT32, 0, x.Name
-		case "bool":
-			return desc.FieldDescriptorProto_TYPE_BOOL, 0, x.Name
-		default:
-			fullName := "." + pkg.Path() + "." + x.Name
-			// Check if the identifier is an already known 'message' or 'enum.
-			// NOTE: This checks all of the parsed gunk packages for the type.
-			// TODO: There is likely a much more robust way to handle this.
-			for _, p := range l.astPkgs[pkg.Path()] {
-				for _, decl := range p.Decls {
-					gd, ok := decl.(*ast.GenDecl)
-					if !ok {
-						continue
-					}
-					for _, spec := range gd.Specs {
-						ts, ok := spec.(*ast.TypeSpec)
-						if !ok {
-							continue
-						}
-						switch ts.Type.(type) {
-						case *ast.StructType:
-							if ts.Name.Name == x.Name {
-								return desc.FieldDescriptorProto_TYPE_MESSAGE, 0, fullName
-							}
-						case *ast.Ident:
-							if ts.Name.Name == x.Name {
-								return desc.FieldDescriptorProto_TYPE_ENUM, 0, fullName
-							}
-						}
-					}
-				}
+		switch typ.Kind() {
+		case types.String:
+			return desc.FieldDescriptorProto_TYPE_STRING, 0, typ.Name()
+		case types.Int, types.Int32:
+			return desc.FieldDescriptorProto_TYPE_INT32, 0, typ.Name()
+		case types.Bool:
+			return desc.FieldDescriptorProto_TYPE_BOOL, 0, typ.Name()
+		}
+	case *types.Named:
+		fullName := "." + typ.String()
+		switch u := typ.Underlying().(type) {
+		case *types.Basic:
+			switch u.Kind() {
+			case types.Int, types.Int32:
+				return desc.FieldDescriptorProto_TYPE_ENUM, 0, fullName
 			}
-			return 0, 0, ""
-		}
-	case *ast.SelectorExpr:
-		id, ok := x.X.(*ast.Ident)
-		if !ok {
-			break
-		}
-		pkg := l.info.ObjectOf(id).(*types.PkgName).Imported()
-		return l.protoType(x.Sel, pkg)
-	case *ast.ArrayType:
-		// Array's arent handled, only slices. Check if this is an array.
-		if x.Len != nil {
-			return 0, 0, ""
+		case *types.Struct:
+			return desc.FieldDescriptorProto_TYPE_MESSAGE, 0, fullName
+		default:
+			fmt.Printf("%T %v\n", typ, typ)
```

Comment:
Does leaving this print in help at all?

Reply:
Nope! Just a debugging print. Thanks for spotting.